src/main/java/org/carlspring/cloud/storage/s3fs/S3Path.java

```diff
@@ -735,11 +736,20 @@ public int hashCode()
      */
     private String encode(String uri)
     {
-        // remove special case URI starting with //
-        uri = uri.replace("//", "/");
-        uri = uri.replaceAll(" ", "%20");
-
-        return uri;
+        try
+        {
+            // URL encode all characters, but then convert known allowed characters back
+            return URLEncoder.encode(uri, "UTF-8")
+                    .replace("%3A", ":")
+                    .replace("%2F", "/")
+                    .replace("+", "%20");
```

Comment:
I have the feeling that there is a better way to do this. For `URL`-s there is `java.net.URLDecoder`. Perhaps we could do something similar? I mean -- in your code above, it would only catch three of the cases that should be decoded.

@ptirador , @steve-todorov , @BjoernAkAManf : Any better ideas?

Reply:
Years back when I originally submitted that PR, I took that code directly from the AWS SDK. The replacement back for : and / was because URLEncoder actually produces `application/x-www-form-urlencoded` which isn't exactly a URI but was the closest at the time. And I think there were times when S3Path.encode was called with the full absolute URI that contained the scheme (s3:) and everything else. So undoing that was necessary.

You are right though. There is an easier way. Spring has a [UriUtils](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/util/UriUtils.html) class that does just this. In fact,  [section 5 here](https://www.baeldung.com/java-url-encoding-decoding) says its a common mistake to encode more than just the path and that is where I got the suggestion.  

Is it OK to add a dependency on Spring? If so, I'll redo this PR. It would effectively look like:

`    private String encode(String uri)
    {
        return UriUtils.encode(uri, "UTF-8");
    }
`

Reply:
I've had a look at the code that you've linked back to in the AWS SDK and I see what you mean, but I think it will be quite a bit limited. Sure, it will work for the cases you're currently catching, but while we're doing this, we might as well do it right for all the other cases that need encoding/decoding as well.

We would like to keep this library as minimal, as possible. Adding a dependency to Spring would make it more complicated. We do intend to have Spring-specific code later on, but in a separate "add-on" module (we'll have a "core" one as well).

If we could find a more suitable lightweight library that can handle this sort of encoding/decoding, this would be great!

We do have `com.google.guava:guava:jar:30.1-jre:compile` on the classpath and. as far as I have checked, they do have some URL escaping. I'm not sure if this would be useful. Might be worth a closer look...?

If there really are no lightweight libraries that we can use, we might as well copy the `UriUtils` class to this project, or simply re-implement it ourseves. (Spring is under an Apache 2.0 license and so is this library and as long as we preserve the original licensing information and credit the authors, we should be okay). It would be better to find an simpler alternative, though, and I'm sure we can find some, if we dig around. :)

Reply:
@elerch , @markschreiber : Any thoughts?

Reply:
I analyzed the code paths in UriUtils eventually settling on [HierarchicalUriComponents.encodeUriComponent()](https://github.com/spring-projects/spring-framework/blob/1565f4b83e7c48eeec9dc74f7eb042dce4dbb49a/spring-web/src/main/java/org/springframework/web/util/HierarchicalUriComponents.java#L331) as the thing doing the work. Looks fairly straightforward. I'm happy to do the work, just want to make sure owners agree before I go down that path.

Reply:
If you ask me, this fix should ultimately be applied to the Amazon SDK (v2), so that users of the library (like us) don't have to do this sort of wizardry on our side. I am open to having this method copied over with a disclaimer comment that it was taken from the Spring SDK. The Guava library also contains a similar class ([UrlEscapers](https://github.com/google/guava/blob/master/guava/src/com/google/common/net/UrlEscapers.java], which, from what I can tell, seems more sophisticated. We are already using that library and perhaps it would be neater to use it?

@amarcionek, please, feel free to join our [chat channel](https://chat.carlspring.org/) where we could further discuss this, if you like.

Reply:
The AWS SDK should be here right: https://github.com/aws/aws-sdk-java
Maybe open an Issue there with a potential fix aswell, while applying a Bandage in the Project here, throught the Guava class. Atleast for me, it seems that is easier to contain. However it might be a problem, considering that the UrlEscapers seems to require more manual work, while the Spring version is harder to misuse. Might be wrong though.

Reply:
The link you're pointing to is Amazon SDK v1 and we've migrated to v2.

BTW, @amarcionek, Amazon cut releases of their SDK-s every day, (and we upgrade to these releases almost every day). If you were to fix this on the side of Amazon's SDK, we would get this fix into our codebase as soon as they cut a release that includes it.

Reply:
[This is the correct link](https://github.com/aws/aws-sdk-java-v2) to the repository of Amazon's SDK (v2).